### PR TITLE
Fix game clock skew from device clock differences

### PR DESCRIPTION
**Pull request by ScaleOvenStove**
## Summary

Fixes a bug where players' device clock skew was charged to the game timer, causing the clock to jump ~30 seconds when a peer's event interleaved with yours (typically on first letter typed). The solution establishes a single server-based clock for all game events and measures elapsed time against the server's clock rather than the device's potentially-skewed local clock.

## Key Changes

- **Server timestamp enforcement**: `SocketManager` now overwrites all client-supplied event timestamps with server time (`Date.now()`), ensuring a single authoritative clock for the game. This prevents mixing wall clocks from different devices.

- **Client-side clock offset tracking**: New `src/lib/timing.js` module tracks the offset between server time and local device time:
  - `recordServerTimestamp()` — samples the offset from live socket events and the `join_game` ack
  - `serverNow()` — returns local time adjusted by the offset
  - `getServerTimeOffset()` / `resetServerTimeOffset()` — accessors and test seam

- **Clock component updated**: `Clock.js` now uses `serverNow()` instead of `Date.now()` when measuring elapsed time, since `startTime` and `lastUpdated` are server timestamps.

- **Game page updated**: Solve time calculation in `Game.js` uses `serverNow()` to measure unaccounted elapsed time, preventing skewed local clocks from being baked into recorded solve times.

- **Join game ack includes server time**: `SocketManager.join_game` now returns `{serverTime: Date.now()}` to seed the client's offset before any live event arrives, so a page refresh mid-solve renders the clock against server time.

- **Comprehensive test coverage**: 
  - New `src/lib/__tests__/timing.test.js` validates offset tracking and `serverNow()` behavior
  - New `src/components/Toolbar/__tests__/Clock.test.js` validates clock display with various timing scenarios, including clock skew edge cases
  - Updated `server/__tests__/SocketManager.test.ts` with tests for timestamp overwriting and server time in ack payloads

## Implementation Details

- The offset is module-level state shared by every game in the tab, sampled from live events only (history sync events are legitimately old and would skew the estimate).
- The estimate is short by one-way network latency (~tens of ms), which is negligible compared to device clock skew (seconds to minutes).
- All changes are backwards compatible — older clients that ignore the `serverTime` field in the ack will still work, though they may see clock jumps if peers have skewed clocks.

https://claude.ai/code/session_0196mVoMkRsCAEQfoZQhpqsb

### CODE REVIEW

**Comment by ScaleOvenStove** [comment]
## CI status

Everything that exercises this diff is green: ESLint, Stylelint, Prettier, Tests (Frontend), Tests (Server), TypeCheck (Frontend), TypeCheck (Server), Lockfile.

**Security Audit is red, and it is not caused by this PR.** A newly-published advisory:

```
high  nanoid: custom generators can loop indefinitely when size is zero
      Vulnerable versions  <3.3.17
      Patched versions     >=3.3.17
      Paths                .>stylelint>postcss>nanoid
      GHSA-2v37-7h3g-55p8
```

`pnpm audit` reads only `package.json` and `pnpm-lock.yaml`, and this branch leaves both byte-identical to `master` (`git diff origin/master -- package.json pnpm-lock.yaml` is empty). The same failure reproduces on the untouched dependency tree, so `master` is red on this check too. It's the same class of breakage as #595 — an advisory boundary moving past the existing `pnpm.overrides` pins.

This also cascades: `build` declares `needs: [..., audit]`, so Build shows as skipped rather than run.

The fix is a one-line addition to `pnpm.overrides` in `package.json` (`"nanoid": ">=3.3.17"`) plus the lockfile update. I've left it out of this PR because it's an unrelated dependency change, and because #595 set the precedent of handling advisory bumps in their own PR — which also unblocks `master` and every other open PR at once, rather than only this one. Happy to add it here instead if you'd rather keep it to a single merge.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/timing.js [hunk 1]
```diff
@@ -1 +1,31 @@
...
+// Only live socket events are valid samples: an event from the initial history
+// sync is legitimately old and would look like an enormous offset.
+let serverTimeOffset = 0;
+
+// Records a server-stamped timestamp from a just-received live event. The
+// estimate is short by the one-way network latency (tens of ms), which is far
+// below the clock skew this exists to cancel out.
+export const recordServerTimestamp = (serverTimestamp) => {
+  if (typeof serverTimestamp !== 'number' || !Number.isFinite(serverTimestamp)) return;
+  serverTimeOffset = serverTimestamp - Date.now();
```
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply src/lib/timing.js]: Correct, and it's a bug I introduced. Fixed in 6a5b3e32.

The framing that made it clear: a sample is `serverStamp - Date.now()`, so with a one-way delay `d ≥ 0` and a device clock error `e`, a sample evaluates to `-e - d` while the true offset is `-e`. Every sample is a lower bound on the true offset, biased low by however long the event was in flight — which means "most recent" was the wrong selector, and the delay sources you name (suspended tab, DB write before the broadcast) are just large values of `d`.

So the estimator now keeps the **largest** sample rather than the newest. A late callback produces a smaller sample and loses, which is exactly the rejection you asked for, without needing to classify samples as delayed or not. A 5-minute window bounds it: once the winning sample is stale, the next sample is adopted outright, so a genuine device clock change is still picked up.

Tests cover a late sample not rewinding `serverNow()`, a better sample being adopted immediately, and adaptation after the window.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/SocketManager.ts [hunk 2]
```diff
@@ -286,10 +291,16 @@ class SocketManager {
...
+          // Stamp every event with server time, overwriting whatever the
+          // client sent. The game clock is accumulated from the diff between
+          // consecutive event timestamps (see tick() in lib/reducers/game),
+          // and the client re-sorts history by timestamp, so mixing wall
+          // clocks from different devices charged each player's clock skew to
+          // the timer: a peer whose device was 30s off made the clock jump
+          // ~30s the moment their event interleaved with yours (usually the
+          // first letter typed). One clock for all events removes both the
+          // skew and the out-of-order re-sorting that surfaced it.
+          event.timestamp = Date.now();
```
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply server/SocketManager.ts]: Valid — fixed in 6a5b3e32 by capturing `const receivedAt = Date.now()` at handler entry, before any validation, and stamping with that.

The interleaving is real: the handler is `async`, so the first `await` yields and the next packet from the same socket starts its own handler. `verifiedGids` is only populated after `gameExists()` resolves, so two rapid events both take the uncached path and their durations can differ.

Added a regression test that drives two `updateCell`s at the same cell through concurrently with the first event's `gameExists` lookup made slow. Verified it fails against the previous code — the second event's stamp came out ~50ms *before* the first's — and passes now.

I stayed with a wall-clock read at entry rather than a monotonic counter, since the value has to remain a real timestamp: the client derives its server/local clock offset from it. Same-millisecond ties are fine — `sortedLastIndexBy` keeps equal timestamps in insertion order, which is arrival order.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply server/SocketManager.ts]: The mechanism is real, but I don't think it justifies a migration, so I've deliberately left this one. Reasoning, since it's a judgement call rather than a "no":

**Scope.** It only touches a game *actively being played across the deploy* whose existing history contains future-stamped events — i.e. a participant with a fast clock. It's bounded by that participant's skew, and it self-heals the moment wall-clock time passes the largest future timestamp in that game's history. For the 30s case in the report, that's a ≤30s transient, once, for whoever happens to be mid-solve at deploy.

**Severity of each half.** Stale `updateCell` replay needs the same cell edited both before and after the deploy inside the skew window; the losing value is recoverable by retyping and settles on the next edit. The clock half is capped by `MAX_CLOCK_INCREMENT` (60s), so it's one bounded jump — the same jump today's code produces *continuously* rather than once.

**Cost of the alternatives.** A per-gid monotonic stamp (`max(now, lastTs + 1)`) needs the last persisted ts per gid, which means a DB read on the first event per game plus a cache to keep it off the hot path, and it doesn't actually remove the transient — it propagates the fast client's future timestamp forward instead, so the clock still absorbs the skew once. A client-side monotonic ordering guard would mean never splicing a late event earlier, which contradicts the deliberate "we must support retroactive updates to the event log" behaviour in `HistoryWrapper.addEvent` and changes replay semantics for everyone, permanently, to smooth a one-time transition.

So: a permanent complexity or semantics cost to avoid a bounded, self-healing, one-time transient. Worth noting that the untreated version of this is the bug being fixed — every game today pays it on every interleave, not just at deploy.

Two things that would change my mind, @ScaleOvenStove — say the word and I'll implement either: if you'd rather this be airtight, the per-gid monotonic stamp is the tractable option. Or, cheaper and no code: deploy at a low-traffic hour so few games are mid-solve.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/timing.js [hunk 3]
```diff
@@ -1 +1,65 @@
...
+
+// Records a server-stamped timestamp from a just-received live event or the
+// join_game ack.
+export const recordServerTimestamp = (serverTimestamp) => {
+  if (typeof serverTimestamp !== 'number' || !Number.isFinite(serverTimestamp)) return;
+  const sample = serverTimestamp - Date.now();
+  const now = monotonicNow();
+  // Adopt the first sample, any sample that beats the current estimate, or any
+  // sample at all once the current estimate has gone stale.
+  if (!hasSample || sample > serverTimeOffset || now - sampledAt > OFFSET_WINDOW_MS) {
```
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply src/lib/timing.js]: Right, and a fair catch on my own fix — the expiry branch did reopen the exact hole the previous commit closed, and at the worst moment: a long suspension produces a stale estimate *and* a delayed sample together, so the two conditions coincide rather than being independent. Fixed in b86d0be4.

Lowering the estimate now requires **two consecutive lower samples that agree within 1s**, on top of the staleness check. The reasoning is that a single lower sample is genuinely ambiguous — either the device clock moved, or that one event was delayed — and one sample can't distinguish them. Two can: independent delays rarely agree closely, while two samples taken after a real clock change agree exactly.

Raising the estimate still takes effect on sight, since delay can only bias a sample downward. That matters here, because the usual way a long suspension ends is a socket reconnect, and the `join_game` ack is a fresh undelayed sample that wins immediately — so the realistic recovery path doesn't depend on the downward rule at all.

The tradeoff I accepted: a genuine device clock change now takes up to `OFFSET_WINDOW_MS` plus one corroborating sample to be believed. That's the right side to err on — a delayed callback is common and rewinds a running clock visibly, while a mid-solve clock change is rare.

Four tests: the ten-minute-suspension case from your comment (offset unchanged, `serverNow()` still advancing), recovery via the join ack, disagreeing low samples never accumulating into an adoption, and adaptation once two agreeing samples arrive past the window.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/SocketManager.ts [hunk 4]
```diff
@@ -209,6 +214,12 @@ class SocketManager {
       });
 
       socket.on('game_event', async (message, ack) => {
+        // Captured before any validation, because the checks below await and so
+        // interleave with the next packet from the same socket: whichever
+        // handler clears its awaits first would otherwise claim the earlier
+        // stamp, and two rapid edits to one cell could persist in reverse
+        // order. Reading the clock here makes the stamp reflect arrival order.
+        const receivedAt = Date.now();
```
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply server/SocketManager.ts]: Agreed, and implemented in b86d0be4 — you're right that my previous reply was wrong to wave off equal stamps.

Where my reasoning failed: I checked that `sortedLastIndexBy` inserts after existing equal elements, so equal timestamps keep *client-side arrival order*, and stopped there. But client arrival order is broadcast order, and `addGameEvent` awaits the insert before emitting — so two same-millisecond handlers completing their inserts in either order can invert the broadcast itself. And on a later reload, `ORDER BY ts ASC` with equal `ts` is simply non-deterministic. Both of the paths you named.

`nextEventStamp()` now returns `max(Date.now(), lastStamp + 1)`, so stamps are strictly increasing per process. That makes the invariant this whole PR rests on — history order follows arrival order — hold by construction instead of by luck of the clock.

Two deliberate details:

- **Drift is capped at 1s.** A sustained burst past 1000 events/sec would make `+1` outrun the clock, and the client derives its server/local offset from these stamps, so an unbounded counter would feed it inflated samples. Past a second of drift it falls back to real time and accepts collisions again — graceful degradation under load rather than a slowly-diverging clock.
- **Monotonic per process, not per cluster.** If this ever runs multiple instances, same-millisecond events routed to different processes can still tie. Ordering would then need a sequence the instances share; not worth building for a single-instance deploy, but worth knowing the guarantee's boundary.

Tests: three packets under a frozen clock come out `t`, `t+1`, `t+2`; and a 1500-event burst under a frozen clock stays within the 1s drift cap.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/SocketManager.ts [hunk 5]
```diff
@@ -286,10 +297,16 @@ class SocketManager {
...
+          // Stamp every event with server time, overwriting whatever the
+          // client sent. The game clock is accumulated from the diff between
+          // consecutive event timestamps (see tick() in lib/reducers/game),
+          // and the client re-sorts history by timestamp, so mixing wall
+          // clocks from different devices charged each player's clock skew to
+          // the timer: a peer whose device was 30s off made the clock jump
+          // ~30s the moment their event interleaved with yours (usually the
+          // first letter typed). One clock for all events removes both the
+          // skew and the out-of-order re-sorting that surfaced it.
+          event.timestamp = receivedAt;
```
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply server/SocketManager.ts:345]: The under-count is real and I called it out in the commit message as a deliberate trade. But I'm not going to implement the fix as described, because working through it surfaces a worse problem than the one it solves.

**Why "normalize the queue into server time while preserving its relative intervals" costs more than it's worth.** The intervals in question are values the client chose. To preserve them, the server has to accept a client-supplied timestamp per event and rebase it — the natural design being a per-socket skew estimate (the mirror of the client's estimator: `skew ≈ min(receivedAt - clientTs)` over recent events, then stamp `clientTs + skew`). That reconstructs offline intervals correctly, and it does keep one clock domain.

The problem is what it hands back. Recorded solve time is derived from event spacing, so putting client timestamps back in the stamping path means **the client controls its own recorded solve time again** — and the interesting direction isn't inflation, it's deflation. Submit an offline queue whose timestamps are milliseconds apart and you've claimed a 20-second solve on a Sunday puzzle. This repo persists those times to `puzzle_solves` and aggregates them into per-puzzle stats, so they're comparative data, not just a personal stopwatch.

Flat server stamping is what makes solve times server-measured, and that property is worth more than offline interval fidelity. Note it's a property the codebase didn't have before this PR — client timestamps were authoritative, so a crafted client could already claim any solve time it liked.

**What the under-count actually costs.** Only time spent solving while disconnected, and only in the recorded/replay time — the live displayed clock is unaffected, since it ticks off `serverNow()` locally. Worth noting the clock was never precise across gaps anyway: `MAX_CLOCK_INCREMENT` already collapses any single gap over 60s, online or off. So this widens an existing imprecision rather than introducing a new class of one.

I'd rather under-count a disconnected stretch than make every recorded time forgeable. @ScaleOvenStove — if you'd weight it the other way, the per-socket rebase above is maybe 30 lines and I'll wire it up; the honest framing is "accurate offline times, forgeable solve times" versus "server-measured times, offline stretches under-counted".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/SocketManager.ts [hunk 6]
```diff
@@ -37,10 +37,30 @@ const RESTRICTABLE_EVENT_TYPES: Record<string, RestrictableAction> = Object.assi
...
+  // ordering guarantee this file relies on — history order follows arrival
+  // order — hold by construction rather than by luck of the clock.
+  nextEventStamp(): number {
+    const now = Date.now();
+    const stamp = now > this.lastEventStamp ? now : this.lastEventStamp + 1;
+    // Under a sustained burst past 1000 events/sec the +1s would outrun real
+    // time. Cap the drift at a second and fall back to colliding stamps beyond
+    // that: clients derive their server/local clock offset from these, so the
+    // stamp staying close to real time matters more than the tie-break.
+    this.lastEventStamp = stamp > now + 1000 ? now : stamp;
```
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply server/SocketManager.ts]: You're right, and this one was a self-inflicted regression — the cap I added made things worse than the ties it was layered on top of. Fixed in a8786a7b by removing it.

The tension is real and can't be resolved by a smarter cap: above ~1000 events/sec, strictly increasing stamps *must* drift, so any bound on the drift has to move the sequence backward. Given the choice, monotonicity is the property worth keeping — it's what protects ordering, which is data correctness. Drift costs display precision only, and it decays as soon as the rate drops below one event per millisecond.

I also underrated the trigger, as you say: stamps are claimed before validation, so rejected packets consume sequence too, and the 50/s limit is per socket — so it's a couple of dozen saturating clients, not some theoretical flood.

What I'll note as the residual: sustained overload now inflates stamps ahead of true server time, and clients read their clock offset from those stamps. That's bounded by how long the overload lasts and self-corrects, and normal solving (a few events/sec per player) leaves zero drift. Your "independent ordering key" is the right answer if that ever stops being true — the timestamp would go back to being purely a clock reading and a sequence column would carry ordering. That's a schema change plus a client sort change, so I've left it as the documented escape hatch rather than building it speculatively.

Tests: 3000 stamps under a frozen clock are strictly increasing end to end, and the sequence rejoins real time once it moves past the drift.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/timing.js [hunk 7]
```diff
@@ -1 +1,94 @@
...
+    hasSample = true;
+    candidate = null;
+    return;
+  }
+
+  // Lower than what we have. Believe it only if the previous lower sample
+  // agrees and the estimate we'd be replacing has gone stale.
+  const corroborated = candidate !== null && Math.abs(sample - candidate.value) <= AGREEMENT_TOLERANCE_MS;
+  if (corroborated && now - sampledAt > OFFSET_WINDOW_MS) {
+    serverTimeOffset = sample;
```
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply src/lib/timing.js]: Correct, and this kills the corroboration idea rather than patching it. Fixed in a8786a7b.

The flaw is that I treated agreement as independent evidence when the delay is **common-mode**: a resumed tab drains a batch stamped close together, so their samples land within any tolerance I'd pick. Two observations of the same delay aren't two observations. Tightening the tolerance wouldn't help either — batch-mates stamped 200ms apart agree to 200ms.

Your framing — "freshness or round-trip evidence" — is the actual fix, so the guessing is gone entirely:

- **One-way samples** (broadcast events, unbounded delay) may only **raise** the offset. That direction is safe for a structural reason: delay can only bias a sample low, so a high sample can't be produced by a slow event.
- **Lowering** requires a **measured round trip** — the `join_game` ack, timed on the client. The server stamps it between our send and our receive, so the delay is bounded by the RTT instead of unknown, and the midpoint assumption leaves ±rtt/2 rather than a systematic lag. An RTT too long to bound anything useful (>10s) degrades to raise-only.

Net effect is that the module got smaller: the window, the tolerance, the candidate slot, and the monotonic-clock helper are all gone, replaced by one asymmetry that follows from what a sample actually measures.

The tradeoff I'm accepting: a device clock that moves forward mid-session isn't corrected until the next join — reconnect or refresh. Bounded and self-correcting, and much better than the alternative of a rewound clock corrupting a recorded solve time.

Tests: your batch case (two queued events agreeing after a 10-minute suspension leaves the offset untouched), no quantity of low broadcast samples ever lowering it, the ack correcting downward, the rtt/2 credit, and the >10s-RTT degradation in both directions.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/SocketManager.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply server/SocketManager.ts]: You're right again, and this retracts what I argued in my last reply. Fixed in 302c28b4.

I claimed drift "costs display precision only". That's wrong, and the reason is the one you gave: `tick()` charges `timestamp - clock.lastUpdated` — a *difference*. A constant offset would indeed be harmless; drift **growth** is charged directly as solve time. And since the counter is per process and shared by every game, a burst in one game inflates clocks in all of them, up to `MAX_CLOCK_INCREMENT` per event. That's the same phantom time this PR exists to delete, reintroduced through a different door.

Reachability, checked rather than assumed: the rate limiter is `socket.use` middleware that returns without calling `next()`, so throttled packets never reach the handler and don't claim stamps. But packets that pass the limiter and fail *validation* do, and the limit is 50/s per socket — so ~20 sockets is enough. That matches your read.

The synthetic `+1` is gone. `nextEventStamp()` is now `Date.now()` clamped to never move backward, which keeps the one real win (an NTP step back would otherwise emit stamps into the middle of existing history) and invents no time. Ties are possible again — and to be explicit, that reopens the tie hazard from your earlier comment. Ranked against each other: a tie needs two edits to one cell inside the same millisecond and costs one stale letter that the next edit fixes, whereas drift corrupts recorded solve times across every game on the server. Not a close call.

Your separate-sequence-key suggestion is the way to have both, and I've recorded it at the call site as the fix if ties ever matter — the timestamp stays a pure clock reading and a sequence column carries ordering. I'm not building it in this PR because it's a schema plus client-sort change in service of a hazard I can't construct outside of scripted input.

Tests: 3000 stamps under a frozen clock all equal that instant (nothing invented), and a two-second backward clock step doesn't move the stamp back.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: server/SocketManager.ts [hunk 8]
```diff
@@ -37,10 +37,33 @@ const RESTRICTABLE_EVENT_TYPES: Record<string, RestrictableAction> = Object.assi
...
+  // Deliberately unbounded: past 1000 events/sec in aggregate the +1 outruns
+  // real time and the sequence drifts ahead. Capping that drift would mean
+  // moving the sequence backward, which is far worse than drift — later events
+  // would sort before the preceding second of history and could leave stale
+  // cell values. Drift costs display precision only, needs a sustained rate
+  // this app doesn't see (normal solving is a few events/sec per player), and
+  // decays as soon as the rate drops. If it ever became real, the fix is an
+  // ordering key separate from the timestamp, not a backward reset.
+  nextEventStamp(): number {
+    this.lastEventStamp = Math.max(Date.now(), this.lastEventStamp + 1);
```
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply server/SocketManager.ts]: Accepted — fixed in 302c28b4. Full reasoning is in the sibling thread on the same lines, so briefly here:

You identified the part I got wrong. I defended the unbounded counter as costing "display precision only", but `tick()` charges the *difference* between consecutive stamps, so it's drift growth that gets consumed as solve time — and the counter being per process means one game's burst inflates every game's clock. Reachability checks out too: throttled packets are dropped by `socket.use` before the handler, but packets that pass the limiter and fail validation do claim stamps, so ~20 sockets at 50/s is enough.

The counter is gone. The stamp is `Date.now()` clamped against backward movement and nothing more — no invented milliseconds for `tick()` to bank. Your "separate sequence key" is recorded at the call site as the fix if same-millisecond ties ever prove to matter, deliberately not built here since that's a schema plus client-sort change.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/timing.js [hunk 9]
```diff
@@ -1 +1,74 @@
...
+// evidence that can lower the estimate. Assuming the stamp landed mid-flight
+// puts the residual error at ±roundTripMs / 2.
+export const recordServerTimeFromRoundTrip = (serverTimestamp, roundTripMs) => {
+  if (!isFiniteNumber(serverTimestamp)) return;
+  if (!isFiniteNumber(roundTripMs) || roundTripMs < 0 || roundTripMs > MAX_TRUSTED_ROUND_TRIP_MS) {
+    // Delay isn't bounded after all — treat it as an ordinary one-way sample.
+    recordServerTimestamp(serverTimestamp);
+    return;
+  }
+  serverTimeOffset = serverTimestamp - Date.now() + roundTripMs / 2;
```
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply src/lib/timing.js]: Correct — the midpoint assumption was wrong for this stamp, and your worked example is right. Fixed in 302c28b4.

The flaw: midpoint is the standard correction when the stamp is taken *mid-flight*, but `serverTime` is created after the moderation and ownership lookups, so it sits near the *response*. With 8s of server processing, `+rtt/2` credits ~4s to outbound flight that was actually spent in the handler. And as you note it then sticks, because one-way samples are raise-only — so the error survives until the next reconnect, running the displayed clock and the recorded solve fast.

Rather than model where the stamp sits, `join_game` now reports both ends of its own processing and the client averages each direction separately — the four-timestamp form:

```
sentAt ──────► serverReceivedAt ··· serverSentAt ──────► receivedAt

offset      = ((serverReceivedAt - sentAt) + (serverSentAt - receivedAt)) / 2
networkTrip = (receivedAt - sentAt) - (serverSentAt - serverReceivedAt)
```

Server processing cancels out algebraically rather than being apportioned, so the residual is only the asymmetry between the two network paths. The trust threshold now applies to `networkTrip` rather than total duration, so a slow handler no longer looks like an untrustworthy sample. An older server sending only `serverTime` degrades to a one-way raise-only sample — the split is genuinely unknowable there, so it isn't guessed.

Tests: your 8s-processing case recovers the offset exactly, as does a symmetric 200ms network path; an incoherent exchange (inverted server timestamps) and a 15s one-way trip both degrade to raise-only.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/timing.js [hunk 10]
```diff
@@ -1 +1,93 @@
...
+// Beyond this the network round trip tells us too little about the delay to be
+// worth trusting in the downward direction.
+const MAX_TRUSTED_ROUND_TRIP_MS = 10 * 1000;
+
+// A one-way sample: a server-stamped timestamp off a broadcast event, whose
+// delay is unbounded. Raises the estimate, never lowers it.
+export const recordServerTimestamp = (serverTimestamp) => {
+  if (!isFiniteNumber(serverTimestamp)) return;
+  const sample = serverTimestamp - Date.now();
+  if (!hasSample || sample > serverTimeOffset) {
```
[the remark this answers is not in the document]
Comment (ScaleOvenStove) [reply src/lib/timing.js:43]: Real, and I checked the `isCapped` claim rather than taking it on faith: `Game.js` passes a bare `v2` to the Toolbar, so it's `v2={true}` and `isCapped` is live. An inflated `serverNow()` does make it read as capped — and the two paths disagree, so you get the pause icon (`isPaused` includes `isCapped`) next to an inflated value (`updateClock` reads `props.isPaused`, which doesn't). Both wrong at once.

Partially fixed in 8331ba01, and I want to be precise about which half.

**Bounded (the part that persists):** the recorded solve time. `pages/Game` computed unaccounted elapsed time with no upper bound, so a forward step of any size went straight into `puzzle_solves` — and unlike the display, a written row doesn't self-heal. That computation is now `unaccountedClockTime()` in `lib/timing`, capped at `MAX_CLOCK_INCREMENT`: exactly what `tick()` would have charged had an event arrived instead of the solve landing first. The same bound covers a gap the offset estimate hasn't caught up with, and an offline stretch, so it's a consistency fix as much as a targeted one.

**Not fixed: the display, until the next join.** Your suggested remedy — a periodic timed exchange — is the right one, and I've left it out deliberately rather than dismissed it. It means a recurring socket call plus a per-game interval to start, stop, and tear down. That's new lifecycle in the module where I've now shipped three defects in a row (a drift cap that moved time backward, corroboration that common-mode delay defeated, a midpoint correction for a stamp that isn't at the midpoint), each from adding a mechanism rather than removing one.

I also considered and rejected detecting the step locally by comparing `Date.now()` against `performance.now()`. It looks free, but whether the monotonic clock counts system-suspend time is platform-dependent, so on a platform where it doesn't, an 8-hour laptop sleep reads as an 8-hour forward step and the correction would rewind the clock by 8 hours. Trading a rare bug for a common one.

So the accepted residual is: a forward device-clock step mid-session, on a connection that stays alive, shows a wrong clock until the next reconnect or refresh, with the recorded time now bounded. Reachability is limited — NTP normally slews rather than steps, and the step cases (boot, wake from sleep) break the socket and so trigger the recalibrating join anyway. @ScaleOvenStove has the call on whether that's worth a periodic sync; the pieces are all in place if so, since it would reuse `recordServerTimeExchange` unchanged.

---
_Generated by [Claude Code](https://claude.ai/code)_
